### PR TITLE
summarize total size of model params in bytes

### PR DESCRIPTION
**Pull request by kartik4949**
## What does this PR do?
A simplified version of model size i.e only consider total params for calculating model size.
followup pr of #4810 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [ ] **Check that target branch and milestone match!**

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by Borda** [comment]
@kartik4949 nice enhancement, mind rebase it on `release/1.2-dev`?

File: pytorch_lightning/core/memory.py [hunk 1]
```diff
@@ -372,6 +401,8 @@ def get_gpu_memory_map() -> Dict[str, int]:
     }
     return gpu_memory_map
 
+def get_formatted_model_size(total_model_size: float) -> float:
+    return f"{total_model_size:.3f}"
```
Comment (rohitgr7) [inline pytorch_lightning/core/memory.py:405]:
```suggestion
    return f"{total_model_size:,.3f}"
```

Comment (rohitgr7) [reply pytorch_lightning/core/memory.py:405]: just do have `,`

File: tests/core/test_memory.py [hunk 2]
```diff
@@ -33,6 +33,21 @@ def forward(self, *args, **kwargs):
...
+    def __init__(self):
+        super().__init__()
+        self.layer1 = nn.Linear(10, 100)
+        self.layer2 = nn.Linear(100, 2)
+        self.pre_calculated_model_size = 0.005
+
+    def forward(self, x):
+        x = self.layer1(x)
+        x = self.layer2(x)
+        return x
```
Comment (rohitgr7) [inline tests/core/test_memory.py:48]: maybe use the existing models already present in the file like `MixedDtypeModel` or `UnorderedModel` since `pre_calculated_model_size` is the only new property defined here.

File: pytorch_lightning/core/memory.py [hunk 3]
```diff
@@ -259,11 +286,11 @@ def __str__(self):
         if self._model.example_input_array is not None:
             arrays.append(["In sizes", self.in_sizes])
             arrays.append(["Out sizes", self.out_sizes])
+        total_parameters = self.total_parameters
+        trainable_parameters = self.trainable_parameters
+        model_size = self.model_size()
```
Comment (rohitgr7) [inline pytorch_lightning/core/memory.py:291]: I'd suggest either make the above 2 as methods or make model_size a property too. WDYT?

Comment (kartik4949) [reply pytorch_lightning/core/memory.py:291]: model_size should be property, I agree was about to change it, thanks for pointing it out :)

File: tests/core/test_memory.py (hunk 2 above)
Comment (kartik4949) [reply tests/core/test_memory.py:48]: @rohitgr7 using this because other model have their own distinct property like MixedDtypeModel has mixed dtype, etc

hence just for goog readablity added such. :)

Comment (rohitgr7) [reply tests/core/test_memory.py:48]: yeah, not a problem, just a suggestion.

Comment (kartik4949) [reply tests/core/test_memory.py:48]: @rohitgr7 sure :)

**Comment by kartik4949** [comment]
@rohitgr7 @Borda @justusschock  completed the PR, I think it's ready for merge.  suggestions are welcomed :)

File: pytorch_lightning/core/memory.py [hunk 4]
```diff
@@ -213,6 +216,29 @@ def out_sizes(self) -> List:
...
+        """
+        Estimates total model size i.e total params size in MBs
+        total params size gives model size in accounting total model params.
+
+        NOTE: Currently only Supported total params size.
+
+        Returns:
+            Total estimated model size(MB).
+        """
+        return self.total_parameters * self._precision_megabytes
```
Comment (awaelchli) [inline pytorch_lightning/core/memory.py:240]: can you improve these docs, and merge the two methods together (remove underscore prefix)? There are two incomplete sentences.

File: pytorch_lightning/core/memory.py [hunk 5]
```diff
@@ -213,6 +216,29 @@ def out_sizes(self) -> List:
...
+    @property
+    def trainable_parameters(self) -> int:
+        return sum(p.numel() for p in self._model.parameters() if p.requires_grad)
+
+    @property
+    def model_size(self) -> float:
+        return self._model_size()
+    def _model_size(self) -> float:
+        """
+        Estimates total model size i.e total params size in MBs
```
Comment (awaelchli) [inline pytorch_lightning/core/memory.py:232]:
```suggestion
        Estimates total model size, i.e., total params size in MBs.
```

File: pytorch_lightning/core/memory.py [hunk 6]
```diff
@@ -213,6 +216,29 @@ def out_sizes(self) -> List:
...
+        total params size gives model size in accounting total model params.
+
+        NOTE: Currently only Supported total params size.
```
Comment (awaelchli) [inline pytorch_lightning/core/memory.py:235]: It is unclear what this means. What else should there be supported?

File: pytorch_lightning/core/memory.py [hunk 7]
```diff
@@ -411,3 +441,6 @@ def get_human_readable_count(number: int) -> str:
         return f"{int(number):,d} {labels[index]}"
     else:
         return f"{number:,.1f} {labels[index]}"
+if __name__ == "__main__":
+    import doctest
+    doctest.testmod()
```
Comment (awaelchli) [inline pytorch_lightning/core/memory.py:446]: please remove :)

File: tests/core/test_memory.py [hunk 8]
```diff
@@ -247,3 +262,25 @@ def forward(self, *args, **kwargs):
...
+    pytest.param(ModelSummary.MODE_FULL),
+    pytest.param(ModelSummary.MODE_TOP),
+])
+def test_model_size(mode):
+    """ Test that model size is calculated correctly. """
+    model = PreCalculatedModel()
+    summary = model.summarize(mode=mode)
+    pre_calculated_model_size = torch.tensor(model.pre_calculated_model_size)
+    model_size = torch.tensor(summary.model_size)
+    assert torch.isclose(model_size, pre_calculated_model_size, atol=1e-4)
```
Comment (awaelchli) [inline tests/core/test_memory.py:276]: perhaps it could be easier to just make a 1 megabyte model and assert the exact value here, instead of the approximation :)

**Review by awaelchli — COMMENTED** [review]
good work
just needs a bit more polish I think :))

File: pytorch_lightning/core/memory.py [hunk 9]
```diff
@@ -180,6 +182,7 @@ def __init__(self, model, mode: str = MODE_DEFAULT):
         self._model = model
         self._mode = mode
         self._layer_summary = self.summarize()
+        self._precision_megabytes = (self._model.precision / 8.0) / (1024 ** 2.0) # 1 byte -> 8 bits)
```
[the remark this answers is not in the document]
Comment (kartik4949) [reply pytorch_lightning/core/memory.py:185]: 1024 ** 2 is more accurate ?

File: pytorch_lightning/core/memory.py (hunk 7 above)
Comment (kartik4949) [reply pytorch_lightning/core/memory.py:446]: sorry by mistake :)

File: pytorch_lightning/core/memory.py (hunk 6 above)
Comment (kartik4949) [reply pytorch_lightning/core/memory.py:235]: @awaelchli

File: pytorch_lightning/core/memory.py (hunk 4 above)
Comment (kartik4949) [reply pytorch_lightning/core/memory.py]: @awaelchli sure I can, the reason I was separating the function because in follow up PR, i will be adding the other model sizes too 
i.e forward pass + input size

File: pytorch_lightning/core/memory.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/core/memory.py:185]: no, it's not a matter of "accuracy". It's a matter of definition. 

1 MB (megabyte) = 1000000 bytes
1 MiB (Megabibyte)  = 1024 * 1024 bytes

so which definition do you want to use? The calculation should match def defintion in docs.

File: tests/core/test_memory.py (hunk 8 above)
Comment (kartik4949) [reply tests/core/test_memory.py:276]: @awaelchl nice suggestion :) thanks

File: pytorch_lightning/core/memory.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (kartik4949) [reply pytorch_lightning/core/memory.py:185]: @awaelchli 1MiB

[the remark this answers is not in the document]
Comment (kartik4949) [reply pytorch_lightning/core/memory.py:185]: didnt noticed this, but certainly a good catch, thanks alot @awaelchli  :)

File: pytorch_lightning/core/memory.py (hunk 6 above)
Comment (kartik4949) [reply pytorch_lightning/core/memory.py]: wrote a better doc

File: tests/core/test_memory.py (hunk 8 above)
Comment (kartik4949) [reply tests/core/test_memory.py]: done

File: pytorch_lightning/core/memory.py (hunk 4 above)
Comment (kartik4949) [reply pytorch_lightning/core/memory.py]: @awaelchli _model_size because it will be mostly in private scope as we have given def model_size as public end point via property

**Comment by kartik4949** [comment]
@awaelchli done suggested changes .. please let e know anything else :)

File: tests/core/test_memory.py [hunk 10]
```diff
@@ -247,3 +263,23 @@ def forward(self, *args, **kwargs):
     model.example_input_array = example_input
     summary = model.summarize(mode=mode)
     assert summary.in_sizes == [expected_size]
+
```
Comment (rohitgr7) [inline tests/core/test_memory.py:266]:
```suggestion


```

File: tests/core/test_memory.py [hunk 11]
```diff
@@ -247,3 +263,23 @@ def forward(self, *args, **kwargs):
...
+@pytest.mark.parametrize(['mode'], [
+    pytest.param(ModelSummary.MODE_FULL),
+    pytest.param(ModelSummary.MODE_TOP),
+])
+def test_model_size(mode):
+    """ Test model size is calculated correctly. """
+    model = PreCalculatedModel()
+    summary = model.summarize(mode=mode)
+    assert model.pre_calculated_model_size == summary.model_size
+
```
Comment (rohitgr7) [inline tests/core/test_memory.py:276]:
```suggestion


```
you can install pre-commit to avoid these

File: pytorch_lightning/core/memory.py [hunk 12]
```diff
@@ -213,6 +216,29 @@ def out_sizes(self) -> List:
...
+    def _model_size(self) -> float:
+        """
+        Calculates the total model params size in MBs.
+
+        NOTE: Currently only model params size is supported.
+
+        Returns:
+            Total params size (MB).
+        """
+        return self.total_parameters * self._precision_megabytes
```
Comment (rohitgr7) [inline pytorch_lightning/core/memory.py:240]:
```suggestion
    def model_size(self) -> float:
        return self.total_parameters * self._precision_megabytes
```
any reason for separating these two??

File: pytorch_lightning/core/memory.py [hunk 13]
```diff
@@ -180,6 +182,7 @@ def __init__(self, model, mode: str = MODE_DEFAULT):
         self._model = model
         self._mode = mode
         self._layer_summary = self.summarize()
+        self._precision_megabytes = (self._model.precision / 8.0) * 1e-6 # 1 byte -> 8 bits)
```
Comment (rohitgr7) [inline pytorch_lightning/core/memory.py:185]: why 1e-6 and not 1/1024**2??

File: pytorch_lightning/core/memory.py (hunk 12 above)
Comment (kartik4949) [reply pytorch_lightning/core/memory.py:240]: in future, i will be add input_size and forward pass support hence will need a separate function for better readability.

File: pytorch_lightning/core/memory.py (hunk 13 above)
Comment (kartik4949) [reply pytorch_lightning/core/memory.py:185]: @awaelchli pointed out good point
1 Mb = 1000**2
1 MiB = 1024 ** 2
as we are using MB hence 1e6

**Comment by kartik4949** [comment]
Reason for adding separate _model_size is

model_size in future will have more complexity as
currently model_size = total_params size (MB)

In Future:
Model_size = total_params size (MB) + input size (MB) + forward pass / backward pass size (MB)

hence added _model_size to hide complexity .

File: pytorch_lightning/core/memory.py (hunk 13 above)
Comment (rohitgr7) [reply pytorch_lightning/core/memory.py:185]: Oh TIL :)

File: pytorch_lightning/core/memory.py (hunk 12 above)
Comment (rohitgr7) [reply pytorch_lightning/core/memory.py:240]: Then let's update it then. For now I'd suggest to merge these.

Comment (kartik4949) [reply pytorch_lightning/core/memory.py]: ok sure make sense, will separate it later then :)

File: pytorch_lightning/core/memory.py (hunk 4 above)
Comment (kartik4949) [reply pytorch_lightning/core/memory.py]: done

**Comment by kartik4949** [comment]
@awaelchli @rohitgr7 done with all suggested changes, 
merged the two function, will separate them in follow up PR. :)

**Review by rohitgr7 — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
this looks super clean now. great stuff. please ping me with your follow up that you planned, I'd like to have a close look :)

**Comment by awaelchli** [comment]
just one final request, could you add this to the changelog file? thx

**Comment by kartik4949** [comment]
> just one final request, could you add this to the changelog file? thx

sure

**Comment by rohitgr7** [comment]
> @rohitgr7
> in changelog,
> I should add changes in the unreleased block right?

@kartik4949 yes.

**Comment by kartik4949** [comment]
> > @rohitgr7
> > in changelog,
> > I should add changes in the unreleased block right?
> 
> @kartik4949 yes.

done!

**Review by gianscarpe — APPROVED** [review]
LGTM, nice feature!

File: pytorch_lightning/core/memory.py [hunk 14]
```diff
@@ -213,6 +216,18 @@ def out_sizes(self) -> List:
...
+    @property
+    def total_parameters(self) -> int:
+        return sum(p.numel() for p in self._model.parameters())
+
+    @property
+    def trainable_parameters(self) -> int:
+        return sum(p.numel() for p in self._model.parameters() if p.requires_grad)
+
+    @property
+    def model_size(self) -> float:
```
Comment (tchaton) [inline pytorch_lightning/core/memory.py:228]: What happens if a model has weights in different precision mode ?

Wont' model_size should check the precision of each weights individually ?

**Review by tchaton — APPROVED** [review]
Overall, looks good !

File: pytorch_lightning/core/memory.py (hunk 14 above)
Comment (kartik4949) [reply pytorch_lightning/core/memory.py:228]: Self.precision_bytes is calculated using self._model.precision 
So if model precision changes, model size changes!
:)

Comment (rohitgr7) [reply pytorch_lightning/core/memory.py:228]: @kartik4949 did you check if `model.precision` is assigned to 16 before the summary is called or not in case precision is set to 16 in Trainer?

Comment (rohitgr7) [reply pytorch_lightning/core/memory.py:228]: also can you add a simple test for that too?

Comment (kartik4949) [reply pytorch_lightning/core/memory.py:228]: @rohitgr7 sure

Comment (kartik4949) [reply pytorch_lightning/core/memory.py:228]: @rohitgr7 it wont change until the .fit method is called :)

Comment (kartik4949) [reply pytorch_lightning/core/memory.py:228]: btw you think we should add this test in test_memory?

Comment (rohitgr7) [reply pytorch_lightning/core/memory.py:228]: yes, you can parametrize that test with `precision=32, 16` and add `pytest.skipif(gpu_available)` decorator to make sure it runs on GPU only.

Comment (kartik4949) [reply pytorch_lightning/core/memory.py:228]: I did exactly above with skip if not native amp available too because of fp16  
i used evalmodeltemplate with max epochs 1
should we use the above or improve class precalculatedmodel with training_step and configure @rohitgr7 ?

Comment (kartik4949) [reply pytorch_lightning/core/memory.py:228]: @rohitgr7 just checked model precision is set before summary gets called, so we are good on this.

**Comment by kartik4949** [comment]
added test to check model_size with different precision. as suggested by @rohitgr7.

**Review by SkafteNicki — APPROVED** [review]

**Comment by caleb603** [comment]
CHANGELOG.md

**Review by caleb603 — APPROVED** [review]

**Review by caleb603 — COMMENTED** [review]
Good
